File: backend/app/api/v1/reports.py
"""
Reports API Routes

This module contains the API endpoints for the reporting system.
"""

from datetime import datetime
from typing import List, Optional, Dict, Any
from fastapi import APIRouter, Depends, HTTPException, Query, Path
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_db
from app.core.exceptions import ValidationError, NotFoundError
from app.services.reports import (
    UpsStatusService, RaidStatusService, HeartBeatService,
    ReportTemplateService, ReportGenerationService, ReportsService
)
from app.schemas.reports import (
    UpsStatusCreate, UpsStatusUpdate, UpsStatusResponse, UpsStatusListResponse,
    RaidStatusCreate, RaidStatusUpdate, RaidStatusResponse, RaidStatusListResponse,
    HeartBeatCreate, HeartBeatUpdate, HeartBeatResponse, HeartBeatListResponse,
    ReportTemplateCreate, ReportTemplateUpdate, ReportTemplateResponse, ReportTemplateListResponse,
    ReportGenerationCreate, ReportGenerationResponse, ReportGenerationListResponse,
    OnlineUsersReportQuery, HistoryReportQuery, NewUsersReportQuery,
    TopUsersReportQuery, SystemLogQuery, BatchReportQuery,
    SystemStatusReport, ReportType
)

router = APIRouter()


# =============================================================================
# UPS Status Endpoints
# =============================================================================

@router.post("/ups-status", response_model=UpsStatusResponse)
async def create_ups_status(
    ups_data: UpsStatusCreate,
    db: AsyncSession = Depends(get_db)
):
    """Create new UPS status record"""
    try:
        service = UpsStatusService(db)
        return await service.create_ups_status(ups_data)
    except ValidationError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/ups-status/{ups_id}", response_model=UpsStatusResponse)
async def get_ups_status(
    ups_id: int = Path(..., description="UPS status ID"),
    db: AsyncSession = Depends(get_db)
):
    """Get UPS status by ID"""
    try:
        service = UpsStatusService(db)
        return await service.get_ups_status(ups_id)
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.put("/ups-status/{ups_id}", response_model=UpsStatusResponse)
async def update_ups_status(
    ups_id: int = Path(..., description="UPS status ID"),
    update_data: UpsStatusUpdate = ...,
    db: AsyncSession = Depends(get_db)
):
    """Update UPS status"""
    try:
        service = UpsStatusService(db)
        return await service.update_ups_status(ups_id, update_data)
    except (ValidationError, NotFoundError) as e:
        status_code = 404 if isinstance(e, NotFoundError) else 400
        raise HTTPException(status_code=status_code, detail=str(e))


@router.delete("/ups-status/{ups_id}")
async def delete_ups_status(
    ups_id: int = Path(..., description="UPS status ID"),
    db: AsyncSession = Depends(get_db)
):
    """Delete UPS status"""
    try:
        service = UpsStatusService(db)
        await service.delete_ups_status(ups_id)
        return {"message": "UPS status deleted successfully"}
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/ups-status", response_model=List[UpsStatusResponse])
async def list_ups_status(
    skip: int = Query(0, ge=0, description="Number of records to skip"),
    limit: int = Query(100, ge=1, le=1000,
                       description="Number of records to return"),
    db: AsyncSession = Depends(get_db)
):
    """List all UPS status records"""
    service = UpsStatusService(db)
    return await service.get_all_ups_status(skip=skip, limit=limit)


@router.get("/ups-status/summary")
async def get_ups_summary(db: AsyncSession = Depends(get_db)):
    """Get UPS status summary"""
    service = UpsStatusService(db)
    return await service.get_ups_summary()


# =============================================================================
# RAID Status Endpoints
# =============================================================================

@router.post("/raid-status", response_model=RaidStatusResponse)
async def create_raid_status(
    raid_data: RaidStatusCreate,
    db: AsyncSession = Depends(get_db)
):
    """Create new RAID status record"""
    try:
        service = RaidStatusService(db)
        return await service.create_raid_status(raid_data)
    except ValidationError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/raid-status/{raid_id}", response_model=RaidStatusResponse)
async def get_raid_status(
    raid_id: int = Path(..., description="RAID status ID"),
    db: AsyncSession = Depends(get_db)
):
    """Get RAID status by ID"""
    try:
        service = RaidStatusService(db)
        return await service.get_raid_status(raid_id)
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.put("/raid-status/{raid_id}", response_model=RaidStatusResponse)
async def update_raid_status(
    raid_id: int = Path(..., description="RAID status ID"),
    update_data: RaidStatusUpdate = ...,
    db: AsyncSession = Depends(get_db)
):
    """Update RAID status"""
    try:
        service = RaidStatusService(db)
        return await service.update_raid_status(raid_id, update_data)
    except (ValidationError, NotFoundError) as e:
        status_code = 404 if isinstance(e, NotFoundError) else 400
        raise HTTPException(status_code=status_code, detail=str(e))


@router.delete("/raid-status/{raid_id}")
async def delete_raid_status(
    raid_id: int = Path(..., description="RAID status ID"),
    db: AsyncSession = Depends(get_db)
):
    """Delete RAID status"""
    try:
        service = RaidStatusService(db)
        await service.delete_raid_status(raid_id)
        return {"message": "RAID status deleted successfully"}
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/raid-status", response_model=List[RaidStatusResponse])
async def list_raid_status(
    skip: int = Query(0, ge=0, description="Number of records to skip"),
    limit: int = Query(100, ge=1, le=1000,
                       description="Number of records to return"),
    db: AsyncSession = Depends(get_db)
):
    """List all RAID status records"""
    service = RaidStatusService(db)
    return await service.get_all_raid_status(skip=skip, limit=limit)


@router.get("/raid-status/summary")
async def get_raid_summary(db: AsyncSession = Depends(get_db)):
    """Get RAID status summary"""
    service = RaidStatusService(db)
    return await service.get_raid_summary()


# =============================================================================
# HeartBeat Endpoints
# =============================================================================

@router.post("/heartbeat", response_model=HeartBeatResponse)
async def create_heartbeat(
    heartbeat_data: HeartBeatCreate,
    db: AsyncSession = Depends(get_db)
):
    """Create new heartbeat record"""
    try:
        service = HeartBeatService(db)
        return await service.create_heartbeat(heartbeat_data)
    except ValidationError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/heartbeat/{heartbeat_id}", response_model=HeartBeatResponse)
async def get_heartbeat(
    heartbeat_id: int = Path(..., description="Heartbeat ID"),
    db: AsyncSession = Depends(get_db)
):
    """Get heartbeat by ID"""
    try:
        service = HeartBeatService(db)
        return await service.get_heartbeat(heartbeat_id)
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.put("/heartbeat/{heartbeat_id}", response_model=HeartBeatResponse)
async def update_heartbeat(
    heartbeat_id: int = Path(..., description="Heartbeat ID"),
    update_data: HeartBeatUpdate = ...,
    db: AsyncSession = Depends(get_db)
):
    """Update heartbeat"""
    try:
        service = HeartBeatService(db)
        return await service.update_heartbeat(heartbeat_id, update_data)
    except (ValidationError, NotFoundError) as e:
        status_code = 404 if isinstance(e, NotFoundError) else 400
        raise HTTPException(status_code=status_code, detail=str(e))


@router.delete("/heartbeat/{heartbeat_id}")
async def delete_heartbeat(
    heartbeat_id: int = Path(..., description="Heartbeat ID"),
    db: AsyncSession = Depends(get_db)
):
    """Delete heartbeat"""
    try:
        service = HeartBeatService(db)
        await service.delete_heartbeat(heartbeat_id)
        return {"message": "Heartbeat deleted successfully"}
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/heartbeat", response_model=List[HeartBeatResponse])
async def list_heartbeats(
    skip: int = Query(0, ge=0, description="Number of records to skip"),
    limit: int = Query(100, ge=1, le=1000,
                       description="Number of records to return"),
    db: AsyncSession = Depends(get_db)
):
    """List all heartbeat records"""
    service = HeartBeatService(db)
    return await service.get_all_heartbeats(skip=skip, limit=limit)


@router.get("/heartbeat/summary")
async def get_heartbeat_summary(db: AsyncSession = Depends(get_db)):
    """Get heartbeat status summary"""
    service = HeartBeatService(db)
    return await service.get_heartbeat_summary()


# =============================================================================
# Report Template Endpoints
# =============================================================================

@router.post("/templates", response_model=ReportTemplateResponse)
async def create_report_template(
    template_data: ReportTemplateCreate,
    db: AsyncSession = Depends(get_db)
):
    """Create new report template"""
    try:
        service = ReportTemplateService(db)
        return await service.create_template(template_data)
    except ValidationError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/templates/{template_id}", response_model=ReportTemplateResponse)
async def get_report_template(
    template_id: int = Path(..., description="Template ID"),
    db: AsyncSession = Depends(get_db)
):
    """Get report template by ID"""
    try:
        service = ReportTemplateService(db)
        return await service.get_template(template_id)
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.put("/templates/{template_id}", response_model=ReportTemplateResponse)
async def update_report_template(
    template_id: int = Path(..., description="Template ID"),
    update_data: ReportTemplateUpdate = ...,
    db: AsyncSession = Depends(get_db)
):
    """Update report template"""
    try:
        service = ReportTemplateService(db)
        return await service.update_template(template_id, update_data)
    except (ValidationError, NotFoundError) as e:
        status_code = 404 if isinstance(e, NotFoundError) else 400
        raise HTTPException(status_code=status_code, detail=str(e))


@router.delete("/templates/{template_id}")
async def delete_report_template(
    template_id: int = Path(..., description="Template ID"),
    db: AsyncSession = Depends(get_db)
):
    """Delete report template"""
    try:
        service = ReportTemplateService(db)
        await service.delete_template(template_id)
        return {"message": "Report template deleted successfully"}
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/templates", response_model=List[ReportTemplateResponse])
async def list_report_templates(
    skip: int = Query(0, ge=0, description="Number of records to skip"),
    limit: int = Query(100, ge=1, le=1000,
                       description="Number of records to return"),
    report_type: Optional[ReportType] = Query(
        None, description="Filter by report type"),
    db: AsyncSession = Depends(get_db)
):
    """List report templates"""
    service = ReportTemplateService(db)
    if report_type:
        return await service.get_templates_by_type(report_type)
    return await service.get_all_templates(skip=skip, limit=limit)


# =============================================================================
# Report Generation Endpoints
# =============================================================================

@router.post("/generate", response_model=ReportGenerationResponse)
async def create_report_generation(
    generation_data: ReportGenerationCreate,
    db: AsyncSession = Depends(get_db)
):
    """Create new report generation request"""
    try:
        service = ReportGenerationService(db)
        return await service.create_report_generation(generation_data)
    except ValidationError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/generate/{generation_id}", response_model=ReportGenerationResponse)
async def get_report_generation(
    generation_id: int = Path(..., description="Generation ID"),
    db: AsyncSession = Depends(get_db)
):
    """Get report generation by ID"""
    try:
        service = ReportGenerationService(db)
        return await service.get_report_generation(generation_id)
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/generate/user/{username}", response_model=List[ReportGenerationResponse])
async def get_user_report_generations(
    username: str = Path(..., description="Username"),
    db: AsyncSession = Depends(get_db)
):
    """Get user's report generations"""
    service = ReportGenerationService(db)
    return await service.get_user_reports(username)


@router.get("/generate/pending", response_model=List[ReportGenerationResponse])
async def get_pending_reports(db: AsyncSession = Depends(get_db)):
    """Get pending report generations"""
    service = ReportGenerationService(db)
    return await service.get_pending_reports()


# =============================================================================
# Report Data Endpoints
# =============================================================================

@router.get("/data/online-users")
async def get_online_users_report(
    nas_ip: Optional[str] = Query(None, description="Filter by NAS IP"),
    username: Optional[str] = Query(None, description="Filter by username"),
    session_timeout_min: Optional[int] = Query(
        None, ge=1, description="Session timeout in minutes"),
    db: AsyncSession = Depends(get_db)
):
    """Get online users report"""
    service = ReportsService(db)
    query = OnlineUsersReportQuery(
        nas_ip=nas_ip,
        username=username,
        session_timeout_min=session_timeout_min
    )
    return await service.get_online_users_report(query)


@router.get("/data/history")
async def get_history_report(
    username: Optional[str] = Query(None, description="Filter by username"),
    nas_ip: Optional[str] = Query(None, description="Filter by NAS IP"),
    start_date: Optional[datetime] = Query(None, description="Start date"),
    end_date: Optional[datetime] = Query(None, description="End date"),
    session_time_min: Optional[int] = Query(
        None, ge=0, description="Minimum session time in minutes"),
    db: AsyncSession = Depends(get_db)
):
    """Get history report"""
    service = ReportsService(db)
    query = HistoryReportQuery(
        username=username,
        nas_ip=nas_ip,
        start_date=start_date,
        end_date=end_date,
        session_time_min=session_time_min
    )
    return await service.get_history_report(query)


@router.get("/data/last-connect")
async def get_last_connect_report(
    limit: int = Query(100, ge=1, le=1000,
                       description="Number of records to return"),
    db: AsyncSession = Depends(get_db)
):
    """Get last connect report"""
    service = ReportsService(db)
    return await service.get_last_connect_report(limit=limit)


@router.get("/data/new-users")
async def get_new_users_report(
    start_date: Optional[datetime] = Query(None, description="Start date"),
    end_date: Optional[datetime] = Query(None, description="End date"),
    group_name: Optional[str] = Query(None, description="Filter by group"),
    db: AsyncSession = Depends(get_db)
):
    """Get new users report"""
    service = ReportsService(db)
    query = NewUsersReportQuery(
        start_date=start_date,
        end_date=end_date,
        group_name=group_name
    )
    return await service.get_new_users_report(query)


@router.get("/data/top-users")
async def get_top_users_report(
    start_date: Optional[datetime] = Query(None, description="Start date"),
    end_date: Optional[datetime] = Query(None, description="End date"),
    limit: int = Query(10, ge=1, le=100, description="Number of top users"),
    order_by: str = Query(
        "total_traffic", regex="^(total_traffic|session_time|session_count)$", description="Order by field"),
    db: AsyncSession = Depends(get_db)
):
    """Get top users report"""
    service = ReportsService(db)
    query = TopUsersReportQuery(
        start_date=start_date,
        end_date=end_date,
        limit=limit,
        order_by=order_by
    )
    return await service.get_top_users_report(query)


@router.get("/data/system-logs")
async def get_system_logs_report(
    log_level: Optional[str] = Query(None, description="Filter by log level"),
    logger_name: Optional[str] = Query(
        None, description="Filter by logger name"),
    username: Optional[str] = Query(None, description="Filter by username"),
    start_date: Optional[datetime] = Query(None, description="Start date"),
    end_date: Optional[datetime] = Query(None, description="End date"),
    search_text: Optional[str] = Query(
        None, description="Search in log messages"),
    db: AsyncSession = Depends(get_db)
):
    """Get system logs report"""
    service = ReportsService(db)
    query = SystemLogQuery(
        log_level=log_level,
        logger_name=logger_name,
        username=username,
        start_date=start_date,
        end_date=end_date,
        search_text=search_text
    )
    return await service.get_system_logs_report(query)


@router.get("/data/batch")
async def get_batch_report(
    batch_name: Optional[str] = Query(
        None, description="Filter by batch name"),
    start_date: Optional[datetime] = Query(None, description="Start date"),
    end_date: Optional[datetime] = Query(None, description="End date"),
    db: AsyncSession = Depends(get_db)
):
    """Get batch operations report"""
    service = ReportsService(db)
    query = BatchReportQuery(
        batch_name=batch_name,
        start_date=start_date,
        end_date=end_date
    )
    return await service.get_batch_report(query)


@router.get("/data/system-status")
async def get_system_status_report(db: AsyncSession = Depends(get_db)):
    """Get comprehensive system status report"""
    service = ReportsService(db)
    return await service.get_system_status_report()


@router.get("/dashboard")
async def get_reports_dashboard(db: AsyncSession = Depends(get_db)):
    """Get reports dashboard summary"""
    service = ReportsService(db)
    return await service.get_reports_dashboard()

File: frontend/src/services/reports.ts
/**
 * Reports API Service
 *
 * Service class for handling all reports-related API calls
 */

import type { AxiosResponse } from 'axios'
import { apiService } from './api'
import type {
  // UPS Status types
  UpsStatusCreate,
  UpsStatusUpdate,
  UpsStatusResponse,

  // RAID Status types
  RaidStatusCreate,
  RaidStatusUpdate,
  RaidStatusResponse,

  // HeartBeat types
  HeartBeatCreate,
  HeartBeatUpdate,
  HeartBeatResponse,

  // Report Template types
  ReportTemplateCreate,
  ReportTemplateUpdate,
  ReportTemplateResponse,

  // Report Generation types
  ReportGenerationCreate,
  ReportGenerationResponse,

  // Report Query types
  OnlineUsersReportQuery,
  HistoryReportQuery,
  NewUsersReportQuery,
  TopUsersReportQuery,
  SystemLogQuery,
  BatchReportQuery,

  // Report Data types
  OnlineUserReport,
  HistoryReportItem,
  NewUserReportItem,
  TopUserReportItem,
  SystemLogReportItem,
  BatchReportItem,
  SystemStatusReport,
  ReportsDashboard,

  // Enums
  ReportType,
} from '@/types/reports'

export class ReportsApiService {
  private readonly baseUrl = '/reports'

  // =============================================================================
  // UPS Status API Methods
  // =============================================================================

  async createUpsStatus(data: UpsStatusCreate): Promise<UpsStatusResponse> {
    const response: AxiosResponse<UpsStatusResponse> = await apiService.post(
      `${this.baseUrl}/ups-status`,
      data,
    )
    return response.data
  }

  async getUpsStatus(id: number): Promise<UpsStatusResponse> {
    const response: AxiosResponse<UpsStatusResponse> = await apiService.get(
      `${this.baseUrl}/ups-status/${id}`,
    )
    return response.data
  }

  async updateUpsStatus(id: number, data: UpsStatusUpdate): Promise<UpsStatusResponse> {
    const response: AxiosResponse<UpsStatusResponse> = await apiService.put(
      `${this.baseUrl}/ups-status/${id}`,
      data,
    )
    return response.data
  }

  async deleteUpsStatus(id: number): Promise<void> {
    await apiService.delete(`${this.baseUrl}/ups-status/${id}`)
  }

  async listUpsStatus(params?: { skip?: number; limit?: number }): Promise<UpsStatusResponse[]> {
    const response: AxiosResponse<UpsStatusResponse[]> = await apiService.get(
      `${this.baseUrl}/ups-status`,
      { params },
    )
    return response.data
  }

  async getUpsSummary(): Promise<Record<string, unknown>> {
    const response: AxiosResponse<Record<string, unknown>> = await apiService.get(
      `${this.baseUrl}/ups-status/summary`,
    )
    return response.data
  }

  // =============================================================================
  // RAID Status API Methods
  // =============================================================================

  async createRaidStatus(data: RaidStatusCreate): Promise<RaidStatusResponse> {
    const response: AxiosResponse<RaidStatusResponse> = await apiService.post(
      `${this.baseUrl}/raid-status`,
      data,
    )
    return response.data
  }

  async getRaidStatus(id: number): Promise<RaidStatusResponse> {
    const response: AxiosResponse<RaidStatusResponse> = await apiService.get(
      `${this.baseUrl}/raid-status/${id}`,
    )
    return response.data
  }

  async updateRaidStatus(id: number, data: RaidStatusUpdate): Promise<RaidStatusResponse> {
    const response: AxiosResponse<RaidStatusResponse> = await apiService.put(
      `${this.baseUrl}/raid-status/${id}`,
      data,
    )
    return response.data
  }

  async deleteRaidStatus(id: number): Promise<void> {
    await apiService.delete(`${this.baseUrl}/raid-status/${id}`)
  }

  async listRaidStatus(params?: { skip?: number; limit?: number }): Promise<RaidStatusResponse[]> {
    const response: AxiosResponse<RaidStatusResponse[]> = await apiService.get(
      `${this.baseUrl}/raid-status`,
      { params },
    )
    return response.data
  }

  async getRaidSummary(): Promise<Record<string, unknown>> {
    const response: AxiosResponse<Record<string, unknown>> = await apiService.get(
      `${this.baseUrl}/raid-status/summary`,
    )
    return response.data
  }

  // =============================================================================
  // HeartBeat API Methods
  // =============================================================================

  async createHeartBeat(data: HeartBeatCreate): Promise<HeartBeatResponse> {
    const response: AxiosResponse<HeartBeatResponse> = await apiService.post(
      `${this.baseUrl}/heartbeat`,
      data,
    )
    return response.data
  }

  async getHeartBeat(id: number): Promise<HeartBeatResponse> {
    const response: AxiosResponse<HeartBeatResponse> = await apiService.get(
      `${this.baseUrl}/heartbeat/${id}`,
    )
    return response.data
  }

  async updateHeartBeat(id: number, data: HeartBeatUpdate): Promise<HeartBeatResponse> {
    const response: AxiosResponse<HeartBeatResponse> = await apiService.put(
      `${this.baseUrl}/heartbeat/${id}`,
      data,
    )
    return response.data
  }

  async deleteHeartBeat(id: number): Promise<void> {
    await apiService.delete(`${this.baseUrl}/heartbeat/${id}`)
  }

  async listHeartBeats(params?: { skip?: number; limit?: number }): Promise<HeartBeatResponse[]> {
    const response: AxiosResponse<HeartBeatResponse[]> = await apiService.get(
      `${this.baseUrl}/heartbeat`,
      { params },
    )
    return response.data
  }

  async getHeartBeatSummary(): Promise<Record<string, unknown>> {
    const response: AxiosResponse<Record<string, unknown>> = await apiService.get(
      `${this.baseUrl}/heartbeat/summary`,
    )
    return response.data
  }

  // =============================================================================
  // Report Template API Methods
  // =============================================================================

  async createReportTemplate(data: ReportTemplateCreate): Promise<ReportTemplateResponse> {
    const response: AxiosResponse<ReportTemplateResponse> = await apiService.post(
      `${this.baseUrl}/templates`,
      data,
    )
    return response.data
  }

  async getReportTemplate(id: number): Promise<ReportTemplateResponse> {
    const response: AxiosResponse<ReportTemplateResponse> = await apiService.get(
      `${this.baseUrl}/templates/${id}`,
    )
    return response.data
  }

  async updateReportTemplate(
    id: number,
    data: ReportTemplateUpdate,
  ): Promise<ReportTemplateResponse> {
    const response: AxiosResponse<ReportTemplateResponse> = await apiService.put(
      `${this.baseUrl}/templates/${id}`,
      data,
    )
    return response.data
  }

  async deleteReportTemplate(id: number): Promise<void> {
    await apiService.delete(`${this.baseUrl}/templates/${id}`)
  }

  async listReportTemplates(params?: {
    skip?: number
    limit?: number
    report_type?: ReportType
  }): Promise<ReportTemplateResponse[]> {
    const response: AxiosResponse<ReportTemplateResponse[]> = await apiService.get(
      `${this.baseUrl}/templates`,
      { params },
    )
    return response.data
  }

  // =============================================================================
  // Report Generation API Methods
  // =============================================================================

  async createReportGeneration(data: ReportGenerationCreate): Promise<ReportGenerationResponse> {
    const response: AxiosResponse<ReportGenerationResponse> = await apiService.post(
      `${this.baseUrl}/generate`,
      data,
    )
    return response.data
  }

  async getReportGeneration(id: number): Promise<ReportGenerationResponse> {
    const response: AxiosResponse<ReportGenerationResponse> = await apiService.get(
      `${this.baseUrl}/generate/${id}`,
    )
    return response.data
  }

  async getUserReportGenerations(username: string): Promise<ReportGenerationResponse[]> {
    const response: AxiosResponse<ReportGenerationResponse[]> = await apiService.get(
      `${this.baseUrl}/generate/user/${username}`,
    )
    return response.data
  }

  async getPendingReports(): Promise<ReportGenerationResponse[]> {
    const response: AxiosResponse<ReportGenerationResponse[]> = await apiService.get(
      `${this.baseUrl}/generate/pending`,
    )
    return response.data
  }

  // =============================================================================
  // Report Data API Methods
  // =============================================================================

  async getOnlineUsersReport(query?: OnlineUsersReportQuery): Promise<OnlineUserReport[]> {
    const response: AxiosResponse<OnlineUserReport[]> = await apiService.get(
      `${this.baseUrl}/data/online-users`,
      { params: query },
    )
    return response.data
  }

  async getHistoryReport(query?: HistoryReportQuery): Promise<HistoryReportItem[]> {
    const response: AxiosResponse<HistoryReportItem[]> = await apiService.get(
      `${this.baseUrl}/data/history`,
      { params: query },
    )
    return response.data
  }

  async getLastConnectReport(limit = 100): Promise<Record<string, unknown>[]> {
    const response: AxiosResponse<Record<string, unknown>[]> = await apiService.get(
      `${this.baseUrl}/data/last-connect`,
      { params: { limit } },
    )
    return response.data
  }

  async getNewUsersReport(query?: NewUsersReportQuery): Promise<NewUserReportItem[]> {
    const response: AxiosResponse<NewUserReportItem[]> = await apiService.get(
      `${this.baseUrl}/data/new-users`,
      { params: query },
    )
    return response.data
  }

  async getTopUsersReport(query?: TopUsersReportQuery): Promise<TopUserReportItem[]> {
    const response: AxiosResponse<TopUserReportItem[]> = await apiService.get(
      `${this.baseUrl}/data/top-users`,
      { params: query },
    )
    return response.data
  }

  async getSystemLogsReport(query?: SystemLogQuery): Promise<SystemLogReportItem[]> {
    const response: AxiosResponse<SystemLogReportItem[]> = await apiService.get(
      `${this.baseUrl}/data/system-logs`,
      { params: query },
    )
    return response.data
  }

  async getBatchReport(query?: BatchReportQuery): Promise<BatchReportItem[]> {
    const response: AxiosResponse<BatchReportItem[]> = await apiService.get(
      `${this.baseUrl}/data/batch`,
      { params: query },
    )
    return response.data
  }

  async getSystemStatusReport(): Promise<SystemStatusReport> {
    const response: AxiosResponse<SystemStatusReport> = await apiService.get(
      `${this.baseUrl}/data/system-status`,
    )
    return response.data
  }

  async getReportsDashboard(): Promise<ReportsDashboard> {
    const response: AxiosResponse<ReportsDashboard> = await apiService.get(
      `${this.baseUrl}/dashboard`,
    )
    return response.data
  }

  // =============================================================================
  // Utility Methods
  // =============================================================================

  async exportReport(
    reportType: string,
    format: 'csv' | 'excel' | 'pdf' | 'json',
    query?: Record<string, unknown>,
  ): Promise<Blob> {
    const response = (await apiService.get(`${this.baseUrl}/export/${reportType}`, {
      params: { ...query, format },
      responseType: 'blob',
    })) as AxiosResponse<Blob>
    return response.data
  }

  async downloadReportFile(generationId: number): Promise<Blob> {
    const response = (await apiService.get(`${this.baseUrl}/download/${generationId}`, {
      responseType: 'blob',
    })) as AxiosResponse<Blob>
    return response.data
  }

  // =============================================================================
  // Real-time Methods
  // =============================================================================

  async getRealtimeSystemStatus(): Promise<Record<string, unknown>> {
    const response = (await apiService.get(
      `${this.baseUrl}/realtime/system-status`,
    )) as AxiosResponse<Record<string, unknown>>
    return response.data
  }

  async getRealtimeOnlineUsers(): Promise<OnlineUserReport[]> {
    const response: AxiosResponse<OnlineUserReport[]> = await apiService.get(
      `${this.baseUrl}/realtime/online-users`,
    )
    return response.data
  }

  // =============================================================================
  // Batch Operations
  // =============================================================================

  async bulkUpdateSystemStatus(
    updates: Array<{
      type: 'ups' | 'raid' | 'heartbeat'
      id: number
      data: Record<string, unknown>
    }>,
  ): Promise<Record<string, unknown>> {
    const response = (await apiService.post(`${this.baseUrl}/bulk/system-status`, {
      updates,
    })) as AxiosResponse<Record<string, unknown>>
    return response.data
  }

  async scheduleReportGeneration(data: {
    report_type: ReportType
    schedule: string // cron expression
    parameters?: Record<string, unknown>
  }): Promise<Record<string, unknown>> {
    const response = (await apiService.post(`${this.baseUrl}/schedule`, data)) as AxiosResponse<
      Record<string, unknown>
    >
    return response.data
  }

  // =============================================================================
  // Analytics Methods
  // =============================================================================

  async getReportAnalytics(dateRange?: {
    start: string
    end: string
  }): Promise<Record<string, unknown>> {
    const response = (await apiService.get(`${this.baseUrl}/analytics`, {
      params: dateRange,
    })) as AxiosResponse<Record<string, unknown>>
    return response.data
  }

  async getUsageStatistics(): Promise<Record<string, unknown>> {
    const response = (await apiService.get(`${this.baseUrl}/analytics/usage`)) as AxiosResponse<
      Record<string, unknown>
    >
    return response.data
  }

  async getPerformanceMetrics(): Promise<Record<string, unknown>> {
    const response = (await apiService.get(
      `${this.baseUrl}/analytics/performance`,
    )) as AxiosResponse<Record<string, unknown>>
    return response.data
  }
}

// Create and export singleton instance
export const reportsApiService = new ReportsApiService()
